rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+- Start Date: 2019-06-29
+- RFC PR: (leave this empty)
+- Kibana Issue: https://github.com/elastic/kibana/issues/33779
+
+# Summary
+
+Http Service in New platform should provide the ability to execute some logic in response to an incoming request and send the result of this operation back.
+
+# Basic example
+Declaring a route handler for `/url` endpoint:
+```typescript
+router.get(
+  { path: '/url', ...[otherRouteParameters] },
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit) => {
+   // logic to handle request ...
+   return t.ok(result);
+);
```

Comment:
This may be out of the scope of this RFC.

In previous discussions, we've talked about having more explicit methods for registering routes. For example, we've talked about having a `registerPublicApi` and a `registerInternalApi` methods to distinguish between internal and public APIs.

Do we intend these functions to live on the `HttpSetup` contract and accept a `Router` or for them to live on the `Router` object itself?

Reply:
> In previous discussions, we've talked about having more explicit methods for registering routes. For example, we've talked about having a registerPublicApi and a registerInternalApi methods to distinguish between internal and public APIs.

Where can I find this discussion? I heard about something similar only once https://github.com/elastic/kibana/issues/33775#issuecomment-477532284

I want to structure for myself what considered as `public API` and `internal API`? what is [system api](https://github.com/restrry/kibana/blob/c7f6999e41692161d4a4fbeedc90b41f91504a3f/src/legacy/core_plugins/kibana/server/lib/system_api.js#L28) in this picture?

Reply:
We discussed this offline at GAH Orlando as part of the discussion to provide OpenAPI specs for public APIs.

In my view:
- Public APIs are stable across minor releases, provide an OpenAPI spec, are documented publicly, and are exposed to other plugins via a JS client.
- Internal APIs are specific to a plugin and do not provide any stability guarantees.

It's not immediately obvious to me exactly what a "system API" is right now, but it seems like the same concept as an Internal API. I just think it should be more explicit (via the HTTP route instead of a header).

I'll move this to a separate discussion here: #40623